File: 2021/day01.go
package main

import (
	"fmt"
	"io"
	"os"
	"strconv"
	"strings"
)

const input = `199
200
208
210
200
207
240
269
260
263`

func loadInput() string {
	file, err := os.OpenFile("input/day01.txt", os.O_RDWR, 0644)
	if err != nil {
		fmt.Println("could not open file")
	}
	defer file.Close()
	content, err := io.ReadAll(file)
	return string(content)
}

func ProcessInput(rawInput string) []int {
	measure := strings.Split(rawInput, "\n")

	var output []int
	for _, x := range measure {
		i, err := strconv.Atoi(x)
		if err == nil {
			output = append(output, i)
		} else {
			fmt.Println(err)
		}
	}
	return output
}

func CountIncrease(measure []int) int {
	var acc int
	for i := range measure {
		if measure[i] < measure[i+1] {
			acc++
		}
		if i == len(measure)-2 {
			break
		}
	}
	return acc
}

func PartOne(input []int) {
	count := CountIncrease(input)
	fmt.Println(count)
}

func sum(array [3]int) int {
	result := 0
	for _, v := range array {
		result += v
	}
	return result
}

func PartTwo(input []int) {
	var wins [][3]int

	for i := 1; i < len(input)-2; i++ {
		wins = append(wins, [3]int{input[i], input[i+1], input[i+2]})
	}
	var winSum []int
	for _, w := range wins {
		winSum = append(winSum, sum(w))
	}
	count := CountIncrease(winSum)
	fmt.Println(count)
}

func main() {
	puzzleInput := loadInput()
	processed := ProcessInput(puzzleInput)
	PartOne(processed)
	PartTwo(processed)
}

File: 2021/day01.py

from typing import Iterable

def load_input():
    with open("input/day01.txt") as f:
        return [int(line) for line in f]

def count_increases(measurements: list[int]) -> int:
    acc = 0
    for x, y in zip(measurements, measurements[1:]):
        if x < y:
            acc += 1
    return acc


def create_measurement_windows(measurements: list[int]) -> Iterable[tuple[int, ...]]:
    size = len(measurements)
    indices = [(start, start + 3) for start in range(size) if start + 2 < size]

    for start, end in indices:
        yield tuple(measurements[start:end])


def part_one(input: list[int]) -> int:
    return count_increases(input)


def part_two(input: list[int]) -> int:
    sum_win = [sum(w) for w in create_measurement_windows(input)]
    return count_increases(sum_win)

def main():
    input = load_input()
    print(part_one(input))
    print(part_two(input))


if __name__ == "__main__":
    main()
